examples/utils/migration-tools/src/migrator/migrator.ts

```diff
@@ -7,395 +7,177 @@ import { TypedEventEmitter } from "@fluid-internal/client-utils";
+			// completed before we finish connecting, and in that case we want to avoid doing anything.
+			if (!this.connected) {
+				await new Promise<void>((resolve) => {
+					this.migrationTool.events.once("connected", () => {
+						resolve();
+					});
+				});
 			}
 
-			// Ensure another client has not already completed the migration.
-			if (this.migrationState !== "migrating") {
+			// Do nothing if the migration has already completed.
+			if (this.migrationTool.migrationResult !== undefined) {
 				return;
 			}
 
-			await completeTheMigration();
-		};
-
-		this._events.emit("migrating");
-
-		this._migrationP = doTheMigration()
-			.then(() => {
-				// We assume that if we resolved that either the migration was completed or we disconnected.
-				// In either case, we should re-enter the state machine to take the appropriate action.
-				if (this.connected) {
-					// We assume if we are still connected after exiting the loop, then we should be in the "migrated"
-					// state. The following assert validates this assumption.
-					assert(
-						this.currentMigrationTool.newContainerId !== undefined,
-						"newContainerId should be defined",
-					);
-				}
-				this._migrationP = undefined;
-				this.takeAppropriateActionForCurrentMigratable();
-			})
-			.catch(console.error);
-	};
-
-	private readonly ensureLoading = (): void => {
-		// We assume ensureLoading() is called a single time after we reach the "migrated" state.
-
-		if (this._migratedLoadP !== undefined) {
-			return;
-		}
-
-		if (this._migrationP !== undefined) {
-			throw new Error("Cannot start loading the migrated before migration is complete");
-		}
-
-		const migrationTool = this.currentMigrationTool;
-		const acceptedMigration = migrationTool.acceptedMigration;
-		if (acceptedMigration === undefined) {
-			throw new Error("Expect an accepted version before migration starts");
-		}
-
-		const migratedId = migrationTool.newContainerId;
-		if (migratedId === undefined) {
-			throw new Error("Migration ended without a new container being created");
-		}
-
-		const doTheLoad = async (): Promise<void> => {
-			// doTheLoad() should only be called once. It will resolve once we complete loading.
-
-			const migrationSupported = await this.simpleLoader.supportsVersion(
-				acceptedMigration.newVersion,
+			// Load the container to at least the acceptance sequence number and export.  We do this with a
+			// separate container to ensure we don't include any local un-ack'd changes.  Late-arriving messages
+			// may or may not make it into the migrated data, there is no guarantee either way.
+			// TODO: Consider making this a read-only client
+			// TODO: Consider more aggressive checks on whether the migration finished and early disconnect, or an abort signal.
+			const sourceContainer = await this.loadSourceContainerCallback();
+			await waitForAtLeastSequenceNumber(
```

Comment:
Not sure if this is necessary but do you think it's worth asserting the sequence number we end up at is okay?

Reply:
Probably not necessary since the implementation is just above in this file and easy to inspect.  The assert check would just be identical to the `resolve()` check in that function.